test/infrastructure/docker/e2e/docker_test.go

```diff
@@ -0,0 +1,202 @@
+	"sigs.k8s.io/cluster-api/test/framework"
+
+	corev1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha2"
+	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1alpha2"
+)
+
+var _ = Describe("Docker", func() {
+	Describe("Cluster Creation", func() {
+
+		namespace := "default"
+
+		clusterGen := &ClusterGenerator{}
+		nodeGen := &NodeGenerator{}
+
+		Context("One node cluster", func() {
+			It("should create a single node cluster", func() {
+				cluster, infraCluster := clusterGen.GenerateCluster(namespace)
+				singleNode := nodeGen.GenerateNode(cluster.GetName())
+				input := &framework.OneNodeClusterInput{
+					Management:    mgmt,
+					Cluster:       cluster,
+					InfraCluster:  infraCluster,
+					Node:          singleNode,
+					CreateTimeout: 3 * time.Minute,
+				}
+
+				framework.OneNodeCluster(input)
+
+				cleanupInput := &framework.CleanUpInput{
+					Management: mgmt,
+					Cluster:    cluster,
+				}
+
+				framework.CleanUp(cleanupInput)
+			})
+		})
+
+		Context("Multi-node controlplane cluster", func() {
+			It("should create a multi-node controlplane cluster", func() {
+				cluster, infraCluster := clusterGen.GenerateCluster(namespace)
+				nodes := make([]framework.Node, 3)
+				for i := range nodes {
+					nodes[i] = nodeGen.GenerateNode(cluster.Name)
+				}
+
+				input := &framework.MultiNodeControlplaneClusterInput{
+					Management:        mgmt,
+					Cluster:           cluster,
+					InfraCluster:      infraCluster,
+					ControlplaneNodes: nodes,
+					CreateTimeout:     2 * time.Minute,
+				}
+				framework.MultiNodeControlPlaneCluster(input)
+
+				cleanupInput := &framework.CleanUpInput{
+					Management: mgmt,
+					Cluster:    cluster,
+				}
+
+				framework.CleanUp(cleanupInput)
+			})
+		})
+	})
+})
+
+type ClusterGenerator struct {
+	counter int
+}
+
+func (c *ClusterGenerator) GenerateCluster(namespace string) (*corev1.Cluster, *infrav1.DockerCluster) {
+	generatedName := fmt.Sprintf("test-%d", c.counter)
+	c.counter += 1
+
+	infraCluster := &infrav1.DockerCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      generatedName,
+		},
+	}
+
+	cluster := &corev1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      generatedName,
+		},
+		Spec: corev1.ClusterSpec{
+			ClusterNetwork: &corev1.ClusterNetwork{
+				Services: &corev1.NetworkRanges{CIDRBlocks: []string{}},
+				Pods:     &corev1.NetworkRanges{CIDRBlocks: []string{"192.168.0.0/16"}},
+			},
+			InfrastructureRef: &v1.ObjectReference{
+				APIVersion: infrav1.GroupVersion.String(),
+				Kind:       framework.TypeToKind(infraCluster),
+				Namespace:  infraCluster.GetNamespace(),
+				Name:       infraCluster.GetName(),
+			},
+		},
+	}
+	return cluster, infraCluster
+}
+
+type NodeGenerator struct {
+	counter int
+}
+
+func (n *NodeGenerator) GenerateNode(clusterName string) framework.Node {
```

Comment:
Similar here to GenerateCluster, I think we can remove a lot of boilerplate and housekeeping for consumers around generating Machine resources.

Reply:
Yeah, there is certainly room for improvement. I want to see what this looks like with CAPA after the framework refactoring and then we can see if there's a nice pattern to extract